node/cn/tracers/api.go

```diff
@@ -98,22 +98,37 @@ type Backend interface {
 	StateAtTransaction(ctx context.Context, block *types.Block, txIndex int, reexec uint64) (blockchain.Message, vm.BlockContext, vm.TxContext, *state.StateDB, error)
 }
 
-// API is the collection of tracing APIs exposed over the private debugging endpoint.
-type API struct {
+// CommonAPI contains
+// - public methods that change behavior depending on `.unsafeTrace` flag.
+// For instance, TraceTransaction and TraceCall may or may not support custom tracers.
+// - private helper methods such as traceTx
+type CommonAPI struct {
 	backend     Backend
 	unsafeTrace bool
 }
 
-// NewAPIUnsafeDisabled creates a new API definition for the tracing methods of the CN service,
-// only allowing predefined tracers.
-func NewAPIUnsafeDisabled(backend Backend) *API {
-	return &API{backend: backend, unsafeTrace: false}
+// API contains public methods that are considered "safe" to expose in public RPC.
+type API struct {
+	CommonAPI
 }
 
-// NewAPI creates a new API definition for the tracing methods of the CN service,
-// allowing both predefined tracers and Javascript snippet based tracing.
-func NewAPI(backend Backend) *API {
-	return &API{backend: backend, unsafeTrace: true}
+// UnsafeAPI contains public methods that are considered "unsafe" to expose in public RPC.
+type UnsafeAPI struct {
+	CommonAPI
+}
+
+// NewUnsafeAPI creates a new UnsafeAPI definition
+func NewUnsafeAPI(backend Backend, unsafeTrace bool) *UnsafeAPI {
+	return &UnsafeAPI{
+		CommonAPI{backend: backend, unsafeTrace: unsafeTrace},
+	}
+}
+
+// NewAPI creates a new API definition
+func NewAPI(backend Backend, unsafeTrace bool) *API {
+	return &API{
+		CommonAPI{backend: backend, unsafeTrace: unsafeTrace},
+	}
```

Comment:
Why receive the `unsafeTrace` flag?

```suggestion
// NewUnsafeAPI creates a new UnsafeAPI definition
func NewUnsafeAPI(backend Backend) *UnsafeAPI {
	return &UnsafeAPI{
		CommonAPI{backend: backend, unsafeTrace: true},
	}
}

// NewAPI creates a new API definition
func NewAPI(backend Backend) *API {
	return &API{
		CommonAPI{backend: backend, unsafeTrace: false},
	}
```

Then in `node/cn/backend.go`,

```
{ Service: NewAPI() },
{ Service: NewUnsafeAPI(), IPCOnly: DisableUnsafe }
```

Then
1. In HTTP/WS, If DisableUnsafe=true, only Safe API
2. In HTTP/WS, If DisableUnsafe=false, SafeAPI + UnsafeAPI
3. In IPC, SafeAPI + UnsafeAPI

In cases 2 and 3, I thnk TraceTransaction and TraceCall are overwritten with unsafe versions, because `handler.RegisterName` is sequentially called.